packages/postgres-database/tests/projects/test_projects_version_control.py

```diff
@@ -0,0 +1,285 @@
+            writeonce={"checksum"},  # TODO:  all? cannot delete snapshots?
+        )
+
+
+class HeadsOrm(BaseOrm[int]):
+    def __init__(self, connection: SAConnection):
+        super().__init__(
+            projects_vc_heads,
+            connection,
+            writeonce={"repo_id"},
+        )
+
+
+# -------------
+
+
+def eval_checksum(workbench: Dict[str, Any]):
+    # FIXME: prototype
+    block_string = json.dumps(workbench, sort_keys=True).encode("utf-8")
+    raw_hash = hashlib.sha256(block_string)
+    return raw_hash.hexdigest()
+
+
+def eval_snapshot_uuid(repo: RowProxy, commit: RowProxy) -> UUID:
+    assert repo.id == commit.repo_id  # nosec
+    return uuid3(UUID(repo.project_uuid), f"{repo.id}.{commit.snapshot_checksum}")
+
+
+async def add_snapshot(
+    project_wc: RowProxy, checksum: str, repo: RowProxy, conn: SAConnection
+) -> str:
+    snapshot_orm = SnapshotsOrm(conn)
+    snapshot_checksum = checksum
+    row_id = await snapshot_orm.insert(
+        checksum=checksum,
+        content={"workbench": project_wc.workbench, "ui": project_wc.ui},
+    )
+    assert row_id == checksum
+    return checksum
+
+
+async def test_basic_workflow(project: RowProxy, conn: SAConnection):
+
+    # git init
+    async with conn.begin():
+        # create repo
+        repo_orm = ReposOrm(conn)
+        repo_id = await repo_orm.insert(project_uuid=project.uuid)
+        assert repo_id is not None
+
+        repo_orm.pin(rowid=repo_id)
+        repo = await repo_orm.fetch()
+        assert repo
+        assert repo.project_uuid == project.uuid
+        assert repo.project_checksum is None
+        assert repo.created == repo.modified
+
+        # create main branch
+        branches_orm = BranchesOrm(conn)
+        branch_id = await branches_orm.insert(repo_id=repo.id)
+        assert branch_id is not None
+
+        branches_orm.pin(rowid=branch_id)
+        main_branch: Optional[RowProxy] = await branches_orm.fetch()
+        assert main_branch
+        assert main_branch.name == "main", "Expected 'main' as default branch"
+        assert main_branch.head_commit_id is None, "still not assigned"
+        assert main_branch.created == main_branch.modified
+
+        # assign head branch
+        heads_orm = HeadsOrm(conn)
+        await heads_orm.insert(repo_id=repo.id, head_branch_id=branch_id)
+
+        heads_orm.pin(rowid=repo.id)
+        head = await heads_orm.fetch()
+        assert head
+
+    #
+    # create first commit -- TODO: separate tests
+
+    # fetch a *full copy* of the project (WC)
+    repo = await repo_orm.fetch("id project_uuid project_checksum")
+    assert repo
+
+    project_orm = ProjectsOrm(conn).pin(uuid=repo.project_uuid)
+    project_wc = await project_orm.fetch()
+    assert project_wc
+    assert project == project_wc
+
+    # eval checksum
+    checksum = eval_checksum(project_wc.workbench)
+    assert repo.project_checksum != checksum
+
+    # take snapshot <=> git add & commit
```

Comment:
so it is a checkpoint or a snapshot. could we maybe use the same nomenclature in the frontend and backend?

Reply:
the concept of ``checkpoint`` will be build on top of all this. 
It is for the webserver API
At this level I wanted to get as close as possible to git so that we could have more flexibility